Updating Metrics page for 1.4 (#2688)
dbt-labs/docs.getdbt.com: website/docs/docs/build/metrics.md

<VersionBlock firstVersion="1.4">

| Field       | Description                                                 | Example                         | Required? |
|-------------|-------------------------------------------------------------|---------------------------------|-----------|
| name        | A unique identifier for the metric                          | new_customers                   | yes       |
| model       | The dbt model that powers this metric                       | dim_customers                   | yes (no for `derived` metrics)|
| label       | A short for name / label for the metric                     | New Customers                   | yes        |
| description | Long form, human-readable description for the metric        | The number of customers who.... | no        |
| calculation_method | The method of calculation (aggregation or derived) that is applied to the expression  | count_distinct | yes       |
| expression  | The expression to aggregate/calculate over | user_id, cast(user_id as int) | yes       |
| timestamp   | The time-based component of the metric                      | signup_date                     | no       |

> | time_grains | One or more "grains" at which the metric can be evaluated. For more information, see the "Custom Calendar" section.   | [day, week, month, quarter, year]              | no       |

Reviewer: If the only change here is `yes` &rarr; `no` under required, I _think_ you could nestle a `VersionBlock` in here, rather than duplicating the whole table:
```
| timestamp   | The time-based component of the metric                      | signup_date                     | <VersionBlock firstVersion="1.4">no</VersionBlock><VersionBlock lastVersion="1.3">yes</VersionBlock>       |
```

Author: @mirnawong1 would that work? This table is already nested in a version block because we've got previous versions for the change to `type` and `sql`. For some reason I've got it in my head that version blocks couldn't be nested within one another.

        value: "'2020-01-01'"
```

### Calendar

> The dbt_metrics package contains a [basic calendar table](/models/dbt_metrics_default_calendar.sql) that is created as part of your `dbt run`. It contains dates between 2010-01-01 and 2029-12-31. 

Reviewer 2: deploy failing due to failed links. changing to the below

```suggestion
The dbt_metrics package contains a [basic calendar table](https://github.com/dbt-labs/dbt_metrics/blob/main/models/dbt_metrics_default_calendar.sql) that is created as part of your `dbt run`. It contains dates between 2010-01-01 and 2029-12-31. 
```

If you want to use a custom calendar, you can replace the default with any table which meets the following requirements:
- Contains a `date_day` column. 
- Contains the following columns: `date_week`, `date_month`, `date_quarter`, `date_year`, or equivalents. 

> - Additional date columns need to be prefixed with `date_`, e.g. `date_4_5_4_month` for a 4-5-4 retail calendar date set. Dimensions can have any name (see [dimensions on calendar tables](#dimensions-on-calendar-tables)).

Reviewer 2: did you mean this github url: https://github.com/dbt-labs/dbt_metrics#dimensions-from-calendar-tables?

```suggestion
- Additional date columns need to be prefixed with `date_`, e.g. `date_4_5_4_month` for a 4-5-4 retail calendar date set. Dimensions can have any name (refer to [dimensions on calendar tables](https://github.com/dbt-labs/dbt_metrics#dimensions-from-calendar-tables).
```

Author: realizing I forgot to include below content. Adding and resolving this conversation

To do this, set the value of the `dbt_metrics_calendar_model` variable in your `dbt_project.yml` file: 
```yaml
#dbt_project.yml
config-version: 2
[...]
vars:
    dbt_metrics_calendar_model: my_custom_calendar
```

<VersionBlock firstVersion="1.3">

> ### Configuration 

Reviewer 2: suggesting a tab item for this to reduce scrolling. i took the liberty of writing it out for you, thoughts? (had to remove the code comments because it wasn't rendering right)

Reviewer 2: ```
<!--tabs for config and project.yml -->

<Tabs>
<TabItem value="config" label="Metric yml">

```yml
# in metrics.yml
version: 2
metrics:
  - name: config_metric
    label: Example Metric with Config
    model: ref('my_model')
    calculation_method: count
    timestamp: date_field
    time_grains: [day, week, month]
    config:
      enabled: True
```

`</TabItem>`

```
<TabItem value="project" label="dbt_project.yml">

```yml
# in dbt_project.yml
metrics: 
  your_project_name: 
    +enabled: true
```

```
</TabItem>
</Tabs>
```

`<!--End of tabs for config and project.yml -->`
```
```

Reviewer 2: ```
<!--tabs for config and project.yml -->

<Tab>
<TabItem value=“config” label=“Metric yml”>
```

```
yml
# in metrics.yml
version: 2
metrics:
  - name: config_metric
    label: Example Metric with Config
    model: ref(‘my_model’)
    calculation_method: count
    timestamp: date_field
    time_grains: [day, week, month]
    config:
      enabled: True
```

```
</TabItem>

<TabItem value=“project” label=“dbt_project.yml”>
```

```
yml
# in dbt_project.yml
metrics: 
  your_project_name: 
    +enabled: true

```

```
</TabItem>
</Tabs>

<!--End of tabs for config and project.yml -->
```

Author: Love it! Implementing this

> Metric nodes now accept `config` dictionaries like other dbt resources. Metric configs can specified in the metric yml itself or for groups of metrics in the `dbt_project.yml` file.

Reviewer 2: ```suggestion
Metric nodes now accept `config` dictionaries like other dbt resources. Specify Metric configs in the metric yml itself, or for groups of metrics in the `dbt_project.yml` file.
```

```yml
# in metrics.yml
version: 2

metrics:
  - name: config_metric
    label: Example Metric with Config
    model: ref('my_model')
    calculation_method: count
    timestamp: date_field
    time_grains: [day, week, month]

    config:
      enabled: True
```

Or:

```yml
# in dbt_project.yml

metrics: 
  your_project_name: 
    +enabled: true
```

#### Accepted Metric Configurations

> Below is the list of metric configs currently accepted.

Reviewer 2: ```suggestion
The following is the list of currently accepted metric configs:
```

| Config | Type | Accepted Values | Default Value | Description |
|--------|------|-----------------|---------------|-------------|
| `enabled` | boolean | True/False | True | Enables or disables a metric node. When disabled, dbt will not consider it as part of your project. |

> | `treat_null_values_as_zero` | boolean | True/False | True | Controls the `coalesce` behavior for metrics. By default, when there are no observations for a metric, the output of the metric as well as period Over period secondary calculations will include a `coalesce({{ field }}, 0)` to return 0's rather than nulls. Setting this config to False instead returns `NULL` values. |

Reviewer 2: suggesting link to period over period for context

```suggestion
| `treat_null_values_as_zero` | boolean | True/False | True | Controls the `coalesce` behavior for metrics. By default, when there are no observations for a metric, the output of the metric as well as [Period over Period](#secondary-calculations) secondary calculations will include a `coalesce({{ field }}, 0)` to return 0's rather than nulls. Setting this config to False instead returns `NULL` values. |
```

| end_date    | `'2022-12-31'` | Limits the date range of data used in the metric claculation by not querying data after this date | Optional |
| where       | `plan='paying_customer'` | A sql statment, or series of sql statements, that alter the **final** CTE in the generated sql. Most often used to limit the data to specific values of dimensions provided | Optional |

### Secondary Calculations
Secondary calculations are window functions you can add to the metric calculation and perform on the primary metric or metrics. 

You can use them to compare values to an earlier period, calculate year-to-date sums, and return rolling averages. You can add custom secondary calculations into dbt projects - for more information on this, reference the [package README](https://github.com/dbt-labs/dbt_metrics#secondary-calculations).

The supported Secondary Calculations are:

#### Period over Period:

> The period over period secondary calculation performs a calculation against the metric(s) in question by either determining the difference or the ratio between two points of time. This other point in time is determined by the input variable which looks at the grain selected in the macro. 

Reviewer 2: switch to active voice

```suggestion
The period over period secondary calculation performs a calculation against the metric(s) in question by either determining the difference or the ratio between two points in time. The input variable, which looks at the grain selected in the macro, determines the other point. 
```

| Input                  | Example | Description | Required |
| -------------------------- | ----------- | ----------- | -----------|
| `comparison_strategy`      | `ratio` or `difference` | How to calculate the delta between the two periods | Yes |
| `interval`                 | 1 | Integer - the number of time grains to look back | Yes |
| `alias`                    | `week_over_week` | The column alias for the resulting calculation | No |
| `metric_list`              | `base_sum_metric` | List of metrics that the secondary calculation should be applied to. Default is all metrics selected | No |

#### Period to Date:

> The period to date secondary calculation performs an aggregation on a defined **period** of time that is equal to or coarser (higher, more aggregated) than the grain selected. Great example of this is when you want to display a month_to_date value alongside your weekly grained metric.

Reviewer 2: suggesting removal of 'coarser' and replace with higher as it reads easier. just a suggestion

```suggestion
The period to date secondary calculation performs an aggregation on a defined period of time that is equal to or higher than the grain selected. For example, when you want to display a month_to_date value alongside your weekly grained metric.
```

| Input                  | Example | Description | Required |
| -------------------------- | ----------- | ----------- | -----------|
| `aggregate`                | `max`, `average` | The aggregation to use in the window function. Options vary based on the primary aggregation and are enforced in [validate_aggregate_coherence()](https://github.com/dbt-labs/dbt_metrics/blob/main/macros/validation/validate_aggregate_coherence.sql). | Yes |
| `period`                   | `"day"`, `"week"` | The time grain to aggregate to. One of [`"day"`, `"week"`, `"month"`, `"quarter"`, `"year"`]. Must be at equal or coarser (higher, more aggregated) granularity than the metric's grain (see [Time Grains](#time-grains) below). In example grain of `month`, the acceptable periods would be `month`, `quarter`, or `year`. | Yes |
| `alias`                    | `month_to_date` | The column alias for the resulting calculation | No |
| `metric_list`              | `base_sum_metric` | List of metrics that the secondary calculation should be applied to. Default is all metrics selected | No |

#### Rolling:

<VersionBlock firstVersion="1.3" >

> The rolling secondary calculation performs an aggregation on a number of rows in metric dataset. For example, if the user selects the `week` grain and sets a rolling secondary calculation to `4` then the value returned will be a rolling 4 week calculation of whatever aggregation type was selected. If the `interval` input is not provided then the rolling caclulation will be unbounded on all preceding rows.

Reviewer 2: ```suggestion
The rolling secondary calculation performs an aggregation on a number of rows in the metric dataset. For example, if the user selects the `week` grain and sets a rolling secondary calculation to `4`, then the value returned will be a rolling 4-week calculation of whatever aggregation type was selected. If the `interval` input is not provided, then the rolling calculation will return unlimited on all preceding rows. 

```

Author: I'm going to resolve this one because unbounded is the more correct terminology.

| Input                      | Example | Description | Required |
| -------------------------- | ----------- | ----------- | -----------|

> | `aggregate`                | `max`, `average` | The aggregation to use in the window function. Options vary based on the primary aggregation and are enforced in [validate_aggregate_coherence()](/macros/secondary_calculations/validate_aggregate_coherence.sql). | Yes |

Reviewer 2: i think [this](https://github.com/dbt-labs/dbt_metrics/blob/main/macros/secondary_calculations/validate_aggregate_coherence.sql) is the correct link but it returns a 'not found' page.

>(/macros/secondary_calculations/validate_aggregate_coherence.sql)

Author: Fixed!

| `interval`                 | 1 | Integer - the number of time grains to look back | No |
| `alias`                    | `month_to_date` | The column alias for the resulting calculation | No |
| `metric_list`              | `base_sum_metric` | List of metrics that the secondary calculation should be applied to. Default is all metrics selected | No |
</VersionBlock>

<VersionBlock lastVersion="1.2" >

> The rolling secondary calculation performs an aggregation on a number of rows in metric dataset. For example, if the user selects the `week` grain and sets a rolling secondary calculation to `4` then the value returned will be a rolling 4 week calculation of whatever aggregation type was selected.

Reviewer 2: ```suggestion
The rolling secondary calculation performs an aggregation on a number of rows in the metric dataset. For example, if the user selects the `week` grain and sets a rolling secondary calculation to `4`, then the value returned will be a rolling 4-week calculation of whatever aggregation type was selected.
```

| Input                      | Example | Description | Required |
| -------------------------- | ----------- | ----------- | -----------|

> | `aggregate`                | `max`, `average` | The aggregation to use in the window function. Options vary based on the primary aggregation and are enforced in [validate_aggregate_coherence()](/macros/secondary_calculations/validate_aggregate_coherence.sql). | Yes |

Reviewer 2: i think [this](https://github.com/dbt-labs/dbt_metrics/blob/main/macros/secondary_calculations/validate_aggregate_coherence.sql) is the correct link but it returns a 'not found' page.

(/macros/secondary_calculations/validate_aggregate_coherence.sq

Author: Fixed!

| `interval`                 | 1 | Integer - the number of time grains to look back | Yes |
| `alias`                    | `month_to_date` | The column alias for the resulting calculation | No |
| `metric_list`              | `base_sum_metric` | List of metrics that the secondary calculation should be applied to. Default is all metrics selected | No |
</VersionBlock>

<VersionBlock firstVersion="1.3" >

#### Prior: 

> The prior secondary calculation returns the value from a specified number of intervals prior to the row. 

Reviewer 2: ```suggestion
The prior secondary calculation returns the value from a specified number of intervals before the row. 
```

```
{% set my_metric_yml -%}
{% raw %}

metrics:
  - name: develop_metric
    model: ref('fact_orders')
    label: Total Discount ($)
    timestamp: order_date
    time_grains: [day, week, month]
    calculation_method: average
    expression: discount_total
    dimensions:
      - had_discount
      - order_country

  - name: derived_metric
    label: Total Discount ($)
    timestamp: order_date
    time_grains: [day, week, month]
    calculation_method: derived
    expression: "{{ metric('develop_metric') }} - 1 "
    dimensions:
      - had_discount
      - order_country

  - name: some_other_metric_not_using
    label: Total Discount ($)
    timestamp: order_date
    time_grains: [day, week, month]
    calculation_method: derived
    expression: "{{ metric('derived_metric') }} - 1 "
    dimensions:
      - had_discount
      - order_country

{% endraw %}
{%- endset %}

select * 
from {{ metrics.develop(
        develop_yml=my_metric_yml,
        metric_list=['derived_metric']
        grain='month'
        )
    }}
```

> The above example will return a dataset that contains the metric provided in the metric list (derived_metric) and the parent metric (develop_metric). It will not contain some_other_metric_not_using as it is not designated in the metric list or a parent of the metrics included.

Reviewer 2: ```suggestion
The above example will return a dataset that contains the metric provided in the metric list (`derived_metric`) and the parent metric (`develop_metric`). It will not contain `some_other_metric_not_using` as it is not designated in the metric list or a parent of the metrics included.
```

> **Important caveat** - The `expression` property for `derived` metrics must be contained within double quotes in order for it to be rendered correctly. 

Reviewer 2: ```suggestion
**Important caveat** - You _must_ wrap the `expression` property for `derived` metrics in double quotes to render it. For example,  `expression: "{{ metric('develop_metric') }} - 1 "`.

```
